keps/sig-storage/20191008-raw-block-support.md

```diff
@@ -0,0 +1,148 @@
+- [Upgrade/Downgrade Strategy](#upgradedowngrade-strategy)
+- [Test Plan](#test-plan)
+  - [Unit tests](#unit-tests)
+  - [E2E tests](#e2e-tests)
+- [Graduation Criteria](#graduation-criteria)
+  - [Alpha -&gt; Beta](#alpha---beta)
+  - [Beta -&gt; GA](#beta---ga)
+- [Implementation History](#implementation-history)
+  - [K8s 1.9: Alpha](#k8s-19-alpha)
+  - [K8s 1.13: Beta](#k8s-113-beta)
+  - [K8s 1.17: GA](#k8s-117-ga)
+<!-- /toc -->
+
+## Summary
+
+This document presents a proposal for managing raw block storage in Kubernetes
+using the persistent volume source API as a consistent model of consumption.
+
+Note that is has been designed & merged before KEP process was introduced:
+https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/raw-block-pv.md
+
+## Motivation
+
+By extending the API for volumes to specifically request a raw block device,
+we provide an explicit method for volume consumption, whereas previously any
+request for storage was always fulfilled with a formatted filesystem, even when
+the underlying storage was block. In addition, the ability to use a raw block
+device without a filesystem will allow Kubernetes better support of high
+performance applications that can utilize raw block devices directly for their
+storage. Block volumes are critical to applications like databases (MongoDB,
+Cassandra) that require consistent I/O performance and low latency. For mission
+critical applications, like SAP, block storage is a requirement.
+
+### Goals
+
+* Enable durable access to block storage
+* Provide flexibility for users/vendors to utilize various types of storage devices
+* Agree on API changes for block
+* Provide a consistent security model for block devices 
+* Provide a means for running containerized block storage offerings as non-privileged container
+
+### Non-Goals
+
+* Support all storage devices natively in upstream Kubernetes. Non-standard storage devices are expected to be managed using extension
+  mechanisms.
+* Provide a means for full integration into the scheduler based on non-storage related requests (CPU, etc.)
+* Provide a means of ensuring specific topology to ensure co-location of the data 
+* CSI volume plugin changes - CSI block volumes are tracked as a separate KEP.
+
+## Proposal
+
+See original design proposal at
+https://raw.githubusercontent.com/kubernetes/community/master/contributors/design-proposals/storage/raw-block-pv.md
+
+## Upgrade/Downgrade Strategy
+
+These situations can happen when various Kubernetes components run with raw block volume feature gate on/off:
+
+* API server on, controller-manager on, kubelet off: kubelet does not see
+  `volumeDevices` section in pods and thus it will run the pods as if the pods
+  did not use the volume at all.
+* API server on, controller-manager off: PV controller will bind / provision
+  filesystem PVs to PVCs that request block volumes.
+* API server off: all PVs / PVCs / Pods that refer to a block volume are
```

Comment:
I think this is not quite true. In the rollback case, if the object already has the field, we allow the field to [continue to exist](https://github.com/kubernetes/kubernetes/blob/34db57b0071aa62f546020ad4d7cb603196dd0d7/pkg/api/pod/util.go#L530).

So we need to make sure we can appropriately handle the fact that the field might be set but feature is off.

Reply:
This results in the previous bullet - PV controller and kubelet ignore the field. I added a note about this